service/history/execution/mutable_state_builder_add_continue_as_new_event_test.go

```diff
@@ -0,0 +1,344 @@
+			},
+			taskgeneratorAffordance: func(taskGenerator *MockMutableStateTaskGenerator, msb *mutableStateBuilder) {
+				taskGenerator.EXPECT().GenerateWorkflowCloseTasks(gomock.Any(), msb.config.WorkflowDeletionJitterRange("domain"))
+			},
+
+			expectedReturnedState:   expectedEndingReturnExecutionState,
+			expectedReturnedHistory: expectedEndingReturnHistoryState,
+		},
+	}
+
+	for name, td := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+
+			msb := &mutableStateBuilder{
+				domainEntry: cache.NewDomainCacheEntryForTest(
+					&persistence.DomainInfo{ID: domainID},
+					&persistence.DomainConfig{},
+					true,
+					&persistence.DomainReplicationConfig{},
+					1,
+					nil,
+				),
+				executionInfo: startingExecutionInfo,
+				logger:        log.NewNoop(),
+				config:        config.NewForTest(),
+			}
+
+			shardContext := shardCtx.NewMockContext(ctrl)
+			historyManager := persistence.NewMockHistoryManager(ctrl)
+			domainCache := cache.NewMockDomainCache(ctrl)
+			taskGenerator := NewMockMutableStateTaskGenerator(ctrl)
+
+			msb.timeSource = clock.NewMockedTimeSourceAt(time.Unix(0, ts3))
+			msb.eventsCache = events.NewCache(shardID,
+				historyManager,
+				config.NewForTest(),
+				log.NewNoop(),
+				metrics.NewNoopMetricsClient(),
+				domainCache)
+			msb.shard = shardContext
+			msb.executionInfo = td.startingState
+			msb.hBuilder = &HistoryBuilder{
+				history:   td.startingHistory,
+				msBuilder: msb,
+			}
+			msb.taskGenerator = taskGenerator
+
+			td.historyManagerAffordance(historyManager)
+			td.domainCacheAffordance(domainCache)
+			td.taskgeneratorAffordance(taskGenerator, msb)
+			td.shardManagerAffordance(shardContext, msb, domainCache)
+
+			_, returnedBuilder, err := msb.AddContinueAsNewEvent(context.Background(),
+				firstEventID,
+				decisionCompletedEventID,
+				"",
+				&types.ContinueAsNewWorkflowExecutionDecisionAttributes{
+					ExecutionStartToCloseTimeoutSeconds: common.Int32Ptr(60),
+					WorkflowType: &types.WorkflowType{
+						Name: "helloWorldWorkflow",
+					},
+					TaskList: &types.TaskList{
+						Name: "helloWorldGroup",
+					},
+					Input: []uint8{110, 117, 108, 108, 10},
+				})
+			resultExecutionInfo := returnedBuilder.GetExecutionInfo()
+
+			if td.expectedErr != nil {
+				assert.ErrorAs(t, err, &td.expectedErr)
+			}
+
+			// these are generated nondeterministically, with a plain guid generator
+			// todo(david): make this mockable
+			td.expectedReturnedState.RunID = "a run id"
+			resultExecutionInfo.RunID = "a run id"
+			td.expectedReturnedState.CreateRequestID = "a request id"
+			resultExecutionInfo.CreateRequestID = "a request id"
+			for _, historyEvent := range returnedBuilder.GetHistoryBuilder().history {
+				if historyEvent.WorkflowExecutionStartedEventAttributes != nil {
+					historyEvent.WorkflowExecutionStartedEventAttributes.OriginalExecutionRunID = "a run id"
+				}
+			}
```

Comment:
another option for such fields that are not predictable to validate in tests is to use `cmpopts.IgnoreFields` and `cmp.Diff` instead of `assert.Equal`. Example https://github.com/uber/cadence/blob/a614c4c1fd1cf94b3722193cec919985304c9acb/common/types/mapper/thrift/shared_test.go#L1407

Reply:
Fair point, I didn't know it had that param.

I would like to refactor the code so that this param can be passed in (I hate when functions use time.Now, uuid generation or math.rand in the middle of their logic) but that can wait.